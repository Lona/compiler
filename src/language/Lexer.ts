export type NextAction = { type: 'next'; value: string }

export type PushAction = { type: 'push'; value: string }

export type PopAction = { type: 'pop' }

export type Action = NextAction | PushAction | PopAction

export type LiteralPrintPattern = {
  type: 'literal'
  value: string
}

export type SequencePrintPattern = {
  type: 'sequence'
  value: PrintPattern[]
}

export type IndexReferencePrintPattern = {
  type: 'indexReference'
  value: number
}

export type PrintPattern =
  | LiteralPrintPattern
  | SequencePrintPattern
  | IndexReferencePrintPattern

export type Rule = {
  name: string
  pattern: string
  action?: Action
  discard: boolean
  print: PrintPattern
}

export type StateDefinition = {
  name: string
  rules: Rule[]
}

export type SimpleStateDefinition = {
  name?: string
  rules: Omit<Rule, 'action'>[]
}

export type Token = {
  type: string
  value: string
  groups: string[]
  position: {
    start: number
    end: number
  }
}

function memoize<I, O>(f: (value: I) => O): (value: I) => O {
  const cache: Map<I, O> = new Map()

  return (value: I): O => {
    if (!cache.has(value)) {
      cache.set(value, f(value))
    }

    return cache.get(value)!
  }
}

const getRegExp = memoize(
  (pattern: string): RegExp => new RegExp('^(?:' + pattern + ')')
)

export class Lexer {
  state: string[]
  stateDefinitions: StateDefinition[]

  constructor(
    states: SimpleStateDefinition | StateDefinition[],
    initialState?: string
  ) {
    const normalized: StateDefinition[] =
      states instanceof Array
        ? states
        : [{ ...states, name: states.name ?? 'main' }]
    this.state = [initialState ?? normalized[0].name]
    this.stateDefinitions = normalized
  }

  get currentState() {
    return this.state[this.state.length - 1]
  }

  tokenize(source: string) {
    let length = source.length
    let pos = 0
    let tokens: Token[] = []

    main: while (pos < length) {
      const currentState = this.currentState

      const stateDefinition = this.stateDefinitions.find(
        definition => definition.name === currentState
      )

      if (!stateDefinition) {
        throw new Error(`Invalid lexer state: ${currentState}`)
      }

      const { rules } = stateDefinition

      const slice = source.slice(pos)

      for (let rule of rules) {
        let regExp = getRegExp(rule.pattern)

        const match = slice.match(regExp)

        if (match) {
          const [value, ...groups] = match

          if (value.length === 0 && !rule.action) {
            throw new Error(
              'Lexer stalled: no input consumed and no action taken.'
            )
          }

          if (rule.discard !== true) {
            tokens.push({
              type: rule.name,
              value,
              groups,
              position: {
                start: pos,
                end: pos + value.length,
              },
            })
          }

          if (rule.action) {
            switch (rule.action.type) {
              case 'next':
                this.state[this.state.length - 1] = rule.action.value
                break
              case 'push':
                this.state.push(rule.action.value)
                break
              case 'pop':
                this.state.pop()
                break
            }
          }

          pos += value.length

          continue main
        }
      }

      console.error(
        `Failed to parse: ${slice}\n`,
        currentState,
        rules.map(rule => rule.pattern)
      )

      break
    }

    return tokens
  }
}

export namespace Builders {
  export function rule(
    name: string,
    options: {
      pattern?: string
      discard?: boolean
      action?: Action
      print?: PrintPattern
    } = {}
  ): Rule {
    const pattern = options.pattern ?? name

    return {
      name,
      pattern,
      discard: options.discard ?? false,
      print: options.print ?? { type: 'literal', value: pattern },
      action: options.action,
    }
  }

  export function literalPrintPattern(value: string): LiteralPrintPattern {
    return { type: 'literal', value }
  }

  export function referencePrintPattern(
    value: number
  ): IndexReferencePrintPattern {
    return { type: 'indexReference', value }
  }

  export function sequencePrintPattern(
    value: PrintPattern[]
  ): SequencePrintPattern {
    return { type: 'sequence', value }
  }
}
