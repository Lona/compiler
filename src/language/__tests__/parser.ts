import {
  Parser,
  Reference,
  ReferencePattern,
  SequencePattern,
  OrPattern,
  ManyPattern,
} from '../Parser'
import { Token } from '../Lexer'

function createToken(type: string): Token {
  return { type, value: '', groups: [], position: { start: 0, end: 0 } }
}

it('parses references', () => {
  const tokens = [createToken('hello')]

  const reference: Reference = { type: 'token', value: 'Token.hello' }

  const pattern: ReferencePattern = {
    type: 'reference',
    value: reference,
  }

  const parser = new Parser({
    nodes: [
      {
        type: 'record',
        name: 'Root',
        fields: [],
        pattern,
      },
    ],
  })

  expect(parser.parseReference(reference, tokens)).toMatchSnapshot()

  expect(parser.parsePattern(pattern, tokens)).toMatchSnapshot()
})

it('parses sequences', () => {
  const tokens = [createToken('hello'), createToken('world')]

  const reference1: Reference = { type: 'token', value: 'Token.hello' }
  const reference2: Reference = { type: 'token', value: 'Token.world' }

  const pattern: SequencePattern = {
    type: 'sequence',
    value: [
      { type: 'reference', value: reference1 },
      { type: 'reference', value: reference2 },
    ],
  }

  const parser = new Parser({
    nodes: [
      {
        type: 'record',
        name: 'Root',
        fields: [],
        pattern,
      },
    ],
  })

  expect(parser.parsePattern(pattern, tokens)).toMatchSnapshot()
})

it('parses or', () => {
  const reference1: Reference = { type: 'token', value: 'Token.hello' }
  const reference2: Reference = { type: 'token', value: 'Token.world' }

  const pattern: OrPattern = {
    type: 'or',
    value: [
      { type: 'reference', value: reference1 },
      { type: 'reference', value: reference2 },
    ],
  }

  const parser = new Parser({
    nodes: [
      {
        type: 'record',
        name: 'Root',
        fields: [],
        pattern,
      },
    ],
  })

  expect(parser.parsePattern(pattern, [createToken('hello')])).toMatchSnapshot()
  expect(parser.parsePattern(pattern, [createToken('world')])).toMatchSnapshot()
})

it('parses many', () => {
  const reference: Reference = { type: 'token', value: 'Token.hello' }

  const pattern: ManyPattern = {
    type: 'many',
    value: { type: 'reference', value: reference },
  }

  const parser = new Parser({
    nodes: [
      {
        type: 'record',
        name: 'Root',
        fields: [],
        pattern,
      },
    ],
  })

  expect(parser.parsePattern(pattern, [])).toMatchSnapshot()
  expect(
    parser.parsePattern(pattern, [
      createToken('hello'),
      createToken('hello'),
      createToken('hello'),
    ])
  ).toMatchSnapshot()
})
