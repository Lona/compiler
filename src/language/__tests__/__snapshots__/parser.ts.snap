// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses many 1`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "match": Array [],
    "pattern": Object {
      "type": "many",
      "value": Object {
        "type": "reference",
        "value": Object {
          "type": "token",
          "value": "Token.hello",
        },
      },
    },
  },
}
`;

exports[`parses many 2`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "match": Array [
      Object {
        "match": Object {
          "groups": Array [],
          "position": Object {
            "end": 0,
            "start": 0,
          },
          "type": "hello",
          "value": "",
        },
        "pattern": Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.hello",
          },
        },
      },
      Object {
        "match": Object {
          "groups": Array [],
          "position": Object {
            "end": 0,
            "start": 0,
          },
          "type": "hello",
          "value": "",
        },
        "pattern": Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.hello",
          },
        },
      },
      Object {
        "match": Object {
          "groups": Array [],
          "position": Object {
            "end": 0,
            "start": 0,
          },
          "type": "hello",
          "value": "",
        },
        "pattern": Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.hello",
          },
        },
      },
    ],
    "pattern": Object {
      "type": "many",
      "value": Object {
        "type": "reference",
        "value": Object {
          "type": "token",
          "value": "Token.hello",
        },
      },
    },
  },
}
`;

exports[`parses or 1`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "match": Object {
      "match": Object {
        "groups": Array [],
        "position": Object {
          "end": 0,
          "start": 0,
        },
        "type": "hello",
        "value": "",
      },
      "pattern": Object {
        "type": "reference",
        "value": Object {
          "type": "token",
          "value": "Token.hello",
        },
      },
    },
    "pattern": Object {
      "type": "or",
      "value": Array [
        Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.hello",
          },
        },
        Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.world",
          },
        },
      ],
    },
  },
}
`;

exports[`parses or 2`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "match": Object {
      "match": Object {
        "groups": Array [],
        "position": Object {
          "end": 0,
          "start": 0,
        },
        "type": "world",
        "value": "",
      },
      "pattern": Object {
        "type": "reference",
        "value": Object {
          "type": "token",
          "value": "Token.world",
        },
      },
    },
    "pattern": Object {
      "type": "or",
      "value": Array [
        Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.hello",
          },
        },
        Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.world",
          },
        },
      ],
    },
  },
}
`;

exports[`parses references 1`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "groups": Array [],
    "position": Object {
      "end": 0,
      "start": 0,
    },
    "type": "hello",
    "value": "",
  },
}
`;

exports[`parses references 2`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "match": Object {
      "groups": Array [],
      "position": Object {
        "end": 0,
        "start": 0,
      },
      "type": "hello",
      "value": "",
    },
    "pattern": Object {
      "type": "reference",
      "value": Object {
        "type": "token",
        "value": "Token.hello",
      },
    },
  },
}
`;

exports[`parses sequences 1`] = `
Object {
  "tokens": Array [],
  "type": "success",
  "value": Object {
    "match": Array [
      Object {
        "match": Object {
          "groups": Array [],
          "position": Object {
            "end": 0,
            "start": 0,
          },
          "type": "hello",
          "value": "",
        },
        "pattern": Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.hello",
          },
        },
      },
      Object {
        "match": Object {
          "groups": Array [],
          "position": Object {
            "end": 0,
            "start": 0,
          },
          "type": "world",
          "value": "",
        },
        "pattern": Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.world",
          },
        },
      },
    ],
    "pattern": Object {
      "type": "sequence",
      "value": Array [
        Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.hello",
          },
        },
        Object {
          "type": "reference",
          "value": Object {
            "type": "token",
            "value": "Token.world",
          },
        },
      ],
    },
  },
}
`;
